chain/index/api.go

```diff
@@ -0,0 +1,330 @@
+package index
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+var ErrChainForked = xerrors.New("chain forked")
+
+func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error) {
+	// return an error if the indexer is not started
+	if !si.started {
+		return nil, errors.New("ChainValidateIndex called before indexer start")
+	}
+
+	// return an error if the indexer is closed
+	if si.isClosed() {
+		return nil, errors.New("ChainValidateIndex called on closed indexer")
+	}
+
+	// this API only works for epoch < head because of deferred execution in Filecoin
+	head := si.cs.GetHeaviestTipSet()
+	if epoch >= head.Height() {
+		return nil, xerrors.Errorf("cannot validate index at epoch %d, can only validate at an epoch less than chain head epoch %d", epoch, head.Height())
+	}
+
+	// fetch the tipset at the given epoch on the canonical chain
+	expectedTs, err := si.cs.GetTipsetByHeight(ctx, epoch, nil, true)
```

Comment:
```suggestion
	expectedTs, err := si.cs.GetTipsetByHeight(ctx, epoch, head.Key(), true)
```

saves a mutex lock at a minumum